### ISSUE TITLE
Allow using unchanged `pyproject.toml` between midstream and downstream

### ISSUE THREAD

**Issue by jiridanek** [opendatahub-io/base-containers#77]
Eliminate any need for differences in `pyproject.toml` when using the PyPI and when using the Red Hat Python package index.

Related to the uv option to set pytorch index
* https://github.com/opendatahub-io/base-containers/issues/2

## The Index-Specific Patterns in `pyproject.toml` That Would Differ Between Midstream and Downstream

There are **four distinct patterns** in `pyproject.toml` that tie a build to a specific package index, and would need to be different (or eliminated) to allow a single unchanged `pyproject.toml` to work with both PyPI (midstream) and the Red Hat AIPCC index (downstream).

### Pattern 1: `tensorflow[and-cuda]` — the CUDA extra marker

**Current state on `main`** (`jupyter/tensorflow/ubi9-python-3.12/pyproject.toml` and `runtimes/tensorflow/ubi9-python-3.12/pyproject.toml`):

```10:10:jupyter/tensorflow/ubi9-python-3.12/pyproject.toml
    "tensorflow[and-cuda]~=2.20.0",
```

The `[and-cuda]` extra is a PyPI-specific convention. TensorFlow on PyPI ships CUDA support via the `[and-cuda]` extra, which pulls in `nvidia-*` packages. The AIPCC index ships a different tensorflow wheel that already includes CUDA support (or handles it differently through the base image).

**The AIPCC switch** (commit `b18b7ff93`, on branch `fork/jd/26/02/moreaipcc_odh`, not yet merged to `main`) removes the extra:

```diff
-    "tensorflow[and-cuda]~=2.20.0",
+    "tensorflow~=2.20.0",
```

**Problem**: If `pyproject.toml` says `tensorflow[and-cuda]`, this works on PyPI but fails on the AIPCC index where the `[and-cuda]` extra doesn't exist (or resolves differently). If it says `tensorflow` (without the extra), it works on AIPCC but on PyPI you don't get CUDA support.

### Pattern 2: `torch==2.7.1+cu128` — the CUDA-specific local version specifier

**Current state on `main`** (`jupyter/trustyai/ubi9-python-3.12/pyproject.toml`):

```8:8:jupyter/trustyai/ubi9-python-3.12/pyproject.toml
    "torch==2.7.1+cu128; platform_machine != 'ppc64le' and platform_machine != 's390x'",
```

The `+cu128` suffix is a PyTorch local version label that selects the CUDA 12.8 wheel from the PyTorch index. It does **not** exist on PyPI proper — it only exists in the PyTorch-hosted index (`https://download.pytorch.org/whl/cu128`). The AIPCC index ships `torch==2.9.0` without any local version suffix.

**Before AIPCC** (prior to commit `9b7e7bc41`), `jupyter/pytorch` had:

```toml
"torch==2.7.1+cu128",
"torchvision==0.22.1+cu128",
```

**After AIPCC switch** (commit `9b7e7bc41`, already on `main` for pytorch but **not yet for trustyai**):

```toml
"torch==2.9.0",
"torchvision==0.24.0",
```

Similarly, ROCm pytorch had `torch==2.7.1+rocm6.3` and `torchvision==0.22.1+rocm6.3`, which became plain `torch==2.9.0` and `torchvision==0.24.0` on AIPCC.

**Problem**: The `+cu128` / `+rocm6.3` suffixed versions can only be resolved from the PyTorch custom index, not from PyPI or AIPCC. Removing the suffix makes it work on AIPCC but then you lose the ability to resolve from the PyTorch index with the correct accelerator variant.

### Pattern 3: `[[tool.uv.index]]` for PyTorch — CUDA-version-specific URL

**Current state on `main`** (`jupyter/pytorch/ubi9-python-3.12/pyproject.toml`):

```48:52:jupyter/pytorch/ubi9-python-3.12/pyproject.toml
# https://docs.astral.sh/uv/concepts/indexes/#package-indexes
[[tool.uv.index]]
name = "pytorch-cuda"
url = "https://download.pytorch.org/whl/cu128"
explicit = true
```

And for ROCm (**before** the AIPCC switch, removed in `9b7e7bc41`):

```toml
[[tool.uv.index]]
name = "pytorch-rocm"
url = "https://download.pytorch.org/whl/rocm6.3"
explicit = true
```

With corresponding `[tool.uv.sources]` mappings:

```toml
[tool.uv.sources]
torch = { index = "pytorch-cuda" }
torchvision = { index = "pytorch-cuda" }
```

**Problem**: These embed a specific CUDA version (`cu128`) or ROCm version (`rocm6.3`) directly in `pyproject.toml`. When AIPCC is the default index, these entries are unnecessary and harmful (they point uv to look in the wrong place). When PyPI is the default index, they're essential (PyPI doesn't carry CUDA torch wheels at all). The CUDA version also changes between releases.

**Current inconsistency on `main`**: `jupyter/pytorch` still has the `[[tool.uv.index]]` block for `pytorch-cuda` but **no `[tool.uv.sources]` entries reference it** — it's a dead stanza left over from the AIPCC migration. Meanwhile `jupyter/trustyai` still **actively uses** the `pytorch-cuda` index for `torch` and `triton`.

### Pattern 4: ROCm Python index and direct wheel URL

**Current state on `main`** (`jupyter/rocm/tensorflow/ubi9-python-3.12/pyproject.toml`):

```11:12:jupyter/rocm/tensorflow/ubi9-python-3.12/pyproject.toml
    # TODO(jdanek): consider adding #sha256=814a5e4842b0c92d63c7a0bb4df0baf51ff2db8615535d83fe8958204b840598
    "tensorflow-rocm @ https://repo.radeon.com/rocm/manylinux/rocm-rel-6.4/tensorflow_rocm-2.18.1-cp312-cp312-manylinux_2_28_x86_64.whl",
```

This is a direct URL dependency — a hardcoded link to a specific wheel on AMD's ROCm repository. It's not going through any index at all. The AIPCC index would carry its own tensorflow-rocm wheel.

**Problem**: This URL is specific to a particular ROCm version (6.4), architecture (x86_64), and Python version (cp312). It **cannot** work unchanged on AIPCC. When using AIPCC, this would need to be replaced with a version from the AIPCC ROCm index.

### Summary Table for JIRA

| Pattern | Where (on `main` today) | PyPI / PyTorch index | AIPCC index | Conflict |
|---|---|---|---|---|
| `tensorflow[and-cuda]` | `jupyter/tensorflow`, `runtimes/tensorflow` | Required for CUDA support | Extra doesn't exist, must use `tensorflow` | `pyproject.toml` must differ |
| `torch==2.7.1+cu128` | `jupyter/trustyai` | Required with pytorch-cuda index | Must use `torch==2.9.0` (no suffix) | `pyproject.toml` must differ |
| `torch==2.7.1+rocm6.3` | Removed from main (was in `jupyter/rocm/pytorch`) | Required with pytorch-rocm index | Must use `torch==2.9.0` | Already resolved on main |
| `[[tool.uv.index]] pytorch-cuda` cu128 URL | `jupyter/pytorch`, `jupyter/trustyai` | Essential for CUDA torch wheels | Not needed, AIPCC has its own torch | Dead stanza in pytorch, active in trustyai |
| `[[tool.uv.index]] pytorch-rocm` rocm6.3 URL | Removed from main (was in `jupyter/rocm/pytorch`) | Essential for ROCm torch wheels | Not needed | Already resolved on main |
| `tensorflow-rocm @ https://repo.radeon.com/...` direct URL | `jupyter/rocm/tensorflow`, `runtimes/rocm-tensorflow` | Only way to get ROCm tensorflow | AIPCC carries its own | `pyproject.toml` must differ |

### Key Commits

- **`9b7e7bc41`** — Removed `+cu128`/`+rocm6.3` suffixes and pytorch index for jupyter pytorch (CUDA + ROCm). Switched to AIPCC's plain `torch==2.9.0`.
- **`b18b7ff93`** (not yet merged, on `fork/jd/26/02/moreaipcc_odh`) — Removed `[and-cuda]` from tensorflow, removed `torch==2.7.1+cu128` and pytorch-cuda index from trustyai. The remaining work.
- **`a35eeb123`** — Historical example of ROCm version divergence (`rocm6.2.4` → `rocm6.3`) that was downstream-only before copy-back.

**Comment by LalatenduMohanty** [comment]
@tiran What is your suggestion on this. 

@jiridanek Just to make sure I understand the issue correctly , this issue is not blocking you from using the midstream base-images but you are looking for a long term solution where you want to have pyproject.toml between midstream and downstream.
